File: internal/matching/orderbook.go
package matching

import (
	"sort"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"e-orderbook/internal/models"
)

type PriceLevel struct {
	price decimal.Decimal

	// queue holds resting orders in arrival order. We advance `head` when orders are filled/cancelled.
	queue []*models.Order
	head  int
}

func (pl *PriceLevel) add(order *models.Order) {
	pl.queue = append(pl.queue, order)
}

func (pl *PriceLevel) topValid() *models.Order {
	for pl.head < len(pl.queue) {
		o := pl.queue[pl.head]
		if o.Remaining.LessThanOrEqual(decimal.Zero) || o.Status == models.StatusCancelled || o.Status == models.StatusFilled {
			pl.head++
			continue
		}
		return o
	}

	// Optional compaction for long-running systems.
	if pl.head > 0 {
		pl.queue = nil
		pl.head = 0
	}
	return nil
}

func (pl *PriceLevel) totalRemaining() decimal.Decimal {
	total := decimal.Zero
	for i := pl.head; i < len(pl.queue); i++ {
		o := pl.queue[i]
		if o.Remaining.GreaterThan(decimal.Zero) && (o.Status == models.StatusOpen || o.Status == models.StatusPartiallyFill) {
			total = total.Add(o.Remaining)
		}
	}
	return total
}

type OrderBook struct {
	symbol string
	mu     sync.Mutex

	// For buys: best bid is highest price.
	bidHeap   *priceHeap
	bidLevels map[string]*PriceLevel // key: price.String()

	// For sells: best ask is lowest price.
	askHeap   *priceHeap
	askLevels map[string]*PriceLevel // key: price.String()

	// orders contains only currently resting orders (open/partially_filled), for cancel lookup.
	orders map[uuid.UUID]*models.Order

	seq uint64
}

func NewOrderBook(symbol string) *OrderBook {
	return &OrderBook{
		symbol:    symbol,
		bidHeap:   newMaxPriceHeap(),
		bidLevels: make(map[string]*PriceLevel),
		askHeap:   newMinPriceHeap(),
		askLevels: make(map[string]*PriceLevel),
		orders:    make(map[uuid.UUID]*models.Order),
	}
}

func priceKey(p decimal.Decimal) string { return p.String() }

func (ob *OrderBook) nextSeq() uint64 {
	ob.seq++
	return ob.seq
}

func (ob *OrderBook) bestBid() (decimal.Decimal, bool) {
	for {
		p, ok := ob.bidHeap.Peek()
		if !ok {
			return decimal.Decimal{}, false
		}
		k := priceKey(p)
		pl, exists := ob.bidLevels[k]
		if !exists || pl.totalRemaining().LessThanOrEqual(decimal.Zero) || pl.topValid() == nil {
			// Lazy deletion.
			ob.bidHeap.PopPrice()
			continue
		}
		return p, true
	}
}

func (ob *OrderBook) bestAsk() (decimal.Decimal, bool) {
	for {
		p, ok := ob.askHeap.Peek()
		if !ok {
			return decimal.Decimal{}, false
		}
		k := priceKey(p)
		pl, exists := ob.askLevels[k]
		if !exists || pl.totalRemaining().LessThanOrEqual(decimal.Zero) || pl.topValid() == nil {
			ob.askHeap.PopPrice()
			continue
		}
		return p, true
	}
}

func (ob *OrderBook) addResting(order *models.Order) {
	order.TimePrioritySeq = ob.nextSeq()
	order.Status = models.StatusOpen
	if order.Side == models.SideBuy {
		k := priceKey(*order.Price)
		pl := ob.bidLevels[k]
		if pl == nil {
			pl = &PriceLevel{price: *order.Price}
			ob.bidLevels[k] = pl
			ob.bidHeap.PushPrice(*order.Price)
		}
		pl.add(order)
	} else {
		k := priceKey(*order.Price)
		pl := ob.askLevels[k]
		if pl == nil {
			pl = &PriceLevel{price: *order.Price}
			ob.askLevels[k] = pl
			ob.askHeap.PushPrice(*order.Price)
		}
		pl.add(order)
	}
	ob.orders[order.ID] = order
}

func minDecimal(a, b decimal.Decimal) decimal.Decimal {
	if a.LessThan(b) {
		return a
	}
	return b
}

func touchAffected(affected map[uuid.UUID]*models.Order, o *models.Order) {
	if o != nil {
		affected[o.ID] = o
	}
}

func affectedSlice(affected map[uuid.UUID]*models.Order) []*models.Order {
	out := make([]*models.Order, 0, len(affected))
	for _, o := range affected {
		out = append(out, o)
	}
	return out
}

// Submit attempts to match the incoming order against the book and returns:
// trades, every order struct that was mutated (for persistence), and no error for domain outcomes.
// It assumes ob.symbol matches order.Symbol.
func (ob *OrderBook) Submit(order *models.Order, now time.Time) ([]models.Trade, []*models.Order, error) {
	ob.mu.Lock()
	defer ob.mu.Unlock()

	affected := make(map[uuid.UUID]*models.Order)
	touchAffected(affected, order)

	if order.Symbol != ob.symbol {
		return nil, affectedSlice(affected), nil
	}

	if order.Quantity.LessThanOrEqual(decimal.Zero) {
		order.Status = models.StatusRejected
		order.UpdatedAt = now
		return nil, affectedSlice(affected), nil
	}

	trades := make([]models.Trade, 0)
	executedSomething := false

	switch order.Side {
	case models.SideBuy:
		if order.Type == models.OrderTypeLimit && order.Price == nil {
			order.Status = models.StatusRejected
			order.UpdatedAt = now
			return nil, affectedSlice(affected), nil
		}
		for order.Remaining.GreaterThan(decimal.Zero) {
			bestAsk, ok := ob.bestAsk()
			if !ok {
				break
			}

			// For buy limit, stop if best ask is above our limit.
			if order.Type == models.OrderTypeLimit && bestAsk.GreaterThan(*order.Price) {
				break
			}

			pl := ob.askLevels[priceKey(bestAsk)]
			if pl == nil {
				break
			}
			maker := pl.topValid()
			if maker == nil {
				break
			}

			execQty := minDecimal(order.Remaining, maker.Remaining)
			if execQty.LessThanOrEqual(decimal.Zero) {
				break
			}

			order.Remaining = order.Remaining.Sub(execQty)
			maker.Remaining = maker.Remaining.Sub(execQty)
			order.UpdatedAt = now
			maker.UpdatedAt = now

			if maker.Remaining.LessThanOrEqual(decimal.Zero) {
				maker.Remaining = decimal.Zero
				maker.Status = models.StatusFilled
				delete(ob.orders, maker.ID)
			} else {
				maker.Status = models.StatusPartiallyFill
			}
			touchAffected(affected, maker)

			executedSomething = true

			trades = append(trades, models.Trade{
				ID: uuid.New(),
				// Execution uses maker (resting) price.
				Symbol:      ob.symbol,
				Price:       *maker.Price,
				Quantity:    execQty,
				BuyOrderID:  order.ID,
				SellOrderID: maker.ID,
				ExecutedAt:  now,
			})
		}

		if order.Remaining.LessThanOrEqual(decimal.Zero) {
			order.Remaining = decimal.Zero
			order.Status = models.StatusFilled
		} else {
			if order.Type == models.OrderTypeMarket {
				// Market order does not rest; any remainder is unfilled.
				order.Status = models.StatusPartiallyFill
				order.Remaining = order.Remaining // keep remainder for visibility
			} else {
				if executedSomething {
					order.Status = models.StatusPartiallyFill
				} else {
					order.Status = models.StatusOpen
				}
				ob.addResting(order)
			}
		}

		return trades, affectedSlice(affected), nil

	case models.SideSell:
		if order.Type == models.OrderTypeLimit && order.Price == nil {
			order.Status = models.StatusRejected
			order.UpdatedAt = now
			return nil, affectedSlice(affected), nil
		}
		for order.Remaining.GreaterThan(decimal.Zero) {
			bestBid, ok := ob.bestBid()
			if !ok {
				break
			}

			// For sell limit, stop if best bid is below our limit.
			if order.Type == models.OrderTypeLimit && bestBid.LessThan(*order.Price) {
				break
			}

			pl := ob.bidLevels[priceKey(bestBid)]
			if pl == nil {
				break
			}
			maker := pl.topValid()
			if maker == nil {
				break
			}

			execQty := minDecimal(order.Remaining, maker.Remaining)
			if execQty.LessThanOrEqual(decimal.Zero) {
				break
			}

			order.Remaining = order.Remaining.Sub(execQty)
			maker.Remaining = maker.Remaining.Sub(execQty)
			order.UpdatedAt = now
			maker.UpdatedAt = now

			if maker.Remaining.LessThanOrEqual(decimal.Zero) {
				maker.Remaining = decimal.Zero
				maker.Status = models.StatusFilled
				delete(ob.orders, maker.ID)
			} else {
				maker.Status = models.StatusPartiallyFill
			}
			touchAffected(affected, maker)

			executedSomething = true

			trades = append(trades, models.Trade{
				ID: uuid.New(),
				// Execution uses maker (resting) price.
				Symbol:      ob.symbol,
				Price:       *maker.Price,
				Quantity:    execQty,
				BuyOrderID:  maker.ID,
				SellOrderID: order.ID,
				ExecutedAt:  now,
			})
		}

		if order.Remaining.LessThanOrEqual(decimal.Zero) {
			order.Remaining = decimal.Zero
			order.Status = models.StatusFilled
		} else {
			if order.Type == models.OrderTypeMarket {
				// Market order does not rest; any remainder is unfilled.
				order.Status = models.StatusPartiallyFill
				order.Remaining = order.Remaining
			} else {
				if executedSomething {
					order.Status = models.StatusPartiallyFill
				} else {
					order.Status = models.StatusOpen
				}
				ob.addResting(order)
			}
		}

		return trades, affectedSlice(affected), nil
	default:
		order.Status = models.StatusRejected
		order.UpdatedAt = now
		return nil, affectedSlice(affected), nil
	}
}

func (ob *OrderBook) Cancel(orderID uuid.UUID, now time.Time) (*models.Order, bool) {
	ob.mu.Lock()
	defer ob.mu.Unlock()

	o := ob.orders[orderID]
	if o == nil {
		return nil, false
	}
	if o.Status != models.StatusOpen && o.Status != models.StatusPartiallyFill {
		return nil, false
	}

	o.Status = models.StatusCancelled
	o.Remaining = o.Remaining // keep remainder
	o.UpdatedAt = now
	delete(ob.orders, orderID)

	return o, true
}

type BookLevel struct {
	Price    string `json:"price"`
	Quantity string `json:"quantity"`
}

type BookSnapshot struct {
	Symbol   string      `json:"symbol"`
	Bids     []BookLevel `json:"bids"`
	Asks     []BookLevel `json:"asks"`
	BestBid  *string     `json:"best_bid,omitempty"`
	BestAsk  *string     `json:"best_ask,omitempty"`
	Sequence uint64      `json:"sequence,omitempty"`
}

func orderLevelToBookLevel(price decimal.Decimal, total decimal.Decimal) BookLevel {
	return BookLevel{
		Price:    price.String(),
		Quantity: total.String(),
	}
}

func (ob *OrderBook) Snapshot(depth int) BookSnapshot {
	ob.mu.Lock()
	defer ob.mu.Unlock()

	bidPrices := make([]decimal.Decimal, 0, len(ob.bidLevels))
	for _, pl := range ob.bidLevels {
		if pl.totalRemaining().GreaterThan(decimal.Zero) && pl.topValid() != nil {
			bidPrices = append(bidPrices, pl.price)
		}
	}
	sort.Slice(bidPrices, func(i, j int) bool { return bidPrices[i].GreaterThan(bidPrices[j]) })

	askPrices := make([]decimal.Decimal, 0, len(ob.askLevels))
	for _, pl := range ob.askLevels {
		if pl.totalRemaining().GreaterThan(decimal.Zero) && pl.topValid() != nil {
			askPrices = append(askPrices, pl.price)
		}
	}
	sort.Slice(askPrices, func(i, j int) bool { return askPrices[i].LessThan(askPrices[j]) })

	if depth <= 0 {
		depth = 50
	}

	snap := BookSnapshot{
		Symbol: ob.symbol,
		Bids:   make([]BookLevel, 0, depth),
		Asks:   make([]BookLevel, 0, depth),
	}

	for i := 0; i < len(bidPrices) && i < depth; i++ {
		p := bidPrices[i]
		pl := ob.bidLevels[priceKey(p)]
		if pl == nil {
			continue
		}
		total := pl.totalRemaining()
		if total.LessThanOrEqual(decimal.Zero) {
			continue
		}
		snap.Bids = append(snap.Bids, orderLevelToBookLevel(p, total))
	}

	for i := 0; i < len(askPrices) && i < depth; i++ {
		p := askPrices[i]
		pl := ob.askLevels[priceKey(p)]
		if pl == nil {
			continue
		}
		total := pl.totalRemaining()
		if total.LessThanOrEqual(decimal.Zero) {
			continue
		}
		snap.Asks = append(snap.Asks, orderLevelToBookLevel(p, total))
	}

	if len(snap.Bids) > 0 {
		s := snap.Bids[0].Price
		snap.BestBid = &s
	}
	if len(snap.Asks) > 0 {
		s := snap.Asks[0].Price
		snap.BestAsk = &s
	}

	return snap
}

File: matcher/src/orderbook.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use std::cmp::min;
use std::collections::{BTreeMap, HashMap, VecDeque};
use uuid::Uuid;

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct Order {
    pub id: Uuid,
    pub user_id: String,
    pub symbol: String,
    pub side: String,
    #[serde(rename = "type")]
    pub type_: String,
    pub price: Option<String>,
    pub quantity: String,
    pub remaining: String,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    #[serde(default)]
    pub time_priority_seq: u64,
}

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct Trade {
    pub id: Uuid,
    pub symbol: String,
    pub price: String,
    pub quantity: String,
    pub buy_order_id: Uuid,
    pub sell_order_id: Uuid,
    pub executed_at: DateTime<Utc>,
}

struct PriceLevel {
    queue: VecDeque<Uuid>,
    head: usize,
}

impl PriceLevel {
    fn add(&mut self, id: Uuid) {
        self.queue.push_back(id);
    }

    fn top_valid(&mut self, orders: &HashMap<Uuid, Order>) -> Option<Uuid> {
        while self.head < self.queue.len() {
            let id = self.queue[self.head];
            let o = match orders.get(&id) {
                Some(x) => x,
                None => {
                    self.head += 1;
                    continue;
                }
            };
            let rem = parse_dec(&o.remaining).unwrap_or(Decimal::ZERO);
            if rem <= Decimal::ZERO || o.status == "cancelled" || o.status == "filled" {
                self.head += 1;
                continue;
            }
            return Some(id);
        }
        if self.head > 0 {
            self.queue.clear();
            self.head = 0;
        }
        None
    }

    fn total_remaining(&self, orders: &HashMap<Uuid, Order>) -> Decimal {
        let mut total = Decimal::ZERO;
        for i in self.head..self.queue.len() {
            let id = self.queue[i];
            let Some(o) = orders.get(&id) else { continue };
            if o.status != "open" && o.status != "partial_filled" {
                continue;
            }
            if let Ok(r) = parse_dec(&o.remaining) {
                if r > Decimal::ZERO {
                    total += r;
                }
            }
        }
        total
    }
}

pub struct OrderBook {
    symbol: String,
    bid_levels: BTreeMap<Decimal, PriceLevel>,
    ask_levels: BTreeMap<Decimal, PriceLevel>,
    orders: HashMap<Uuid, Order>,
    seq: u64,
    pub sequence: u64,
}

fn parse_dec(s: &str) -> Result<Decimal, rust_decimal::Error> {
    s.parse()
}

fn dec_str(d: Decimal) -> String {
    d.normalize().to_string()
}

impl OrderBook {
    pub fn new(symbol: String) -> Self {
        Self {
            symbol,
            bid_levels: BTreeMap::new(),
            ask_levels: BTreeMap::new(),
            orders: HashMap::new(),
            seq: 0,
            sequence: 0,
        }
    }

    fn next_seq(&mut self) -> u64 {
        self.seq += 1;
        self.seq
    }

    fn bump_sequence(&mut self) {
        self.sequence += 1;
    }

    fn best_ask_price(&mut self) -> Option<Decimal> {
        loop {
            let p = *self.ask_levels.keys().next()?;
            let pl = self.ask_levels.get_mut(&p)?;
            if pl.total_remaining(&self.orders) <= Decimal::ZERO
                || pl.top_valid(&self.orders).is_none()
            {
                self.ask_levels.remove(&p);
                continue;
            }
            return Some(p);
        }
    }

    fn best_bid_price(&mut self) -> Option<Decimal> {
        loop {
            let p = *self.bid_levels.keys().next_back()?;
            let pl = self.bid_levels.get_mut(&p)?;
            if pl.total_remaining(&self.orders) <= Decimal::ZERO
                || pl.top_valid(&self.orders).is_none()
            {
                self.bid_levels.remove(&p);
                continue;
            }
            return Some(p);
        }
    }

    fn add_resting(&mut self, order: &mut Order) {
        order.time_priority_seq = self.next_seq();
        order.status = "open".to_string();
        let price = parse_dec(order.price.as_ref().expect("limit has price")).unwrap();
        let id = order.id;
        self.orders.insert(id, order.clone());
        if order.side == "buy" {
            let pl = self.bid_levels.entry(price).or_insert_with(|| PriceLevel {
                queue: VecDeque::new(),
                head: 0,
            });
            pl.add(id);
        } else {
            let pl = self.ask_levels.entry(price).or_insert_with(|| PriceLevel {
                queue: VecDeque::new(),
                head: 0,
            });
            pl.add(id);
        }
    }

    pub fn submit(&mut self, order: &mut Order, now: DateTime<Utc>) -> (Vec<Trade>, Vec<Order>) {
        let mut affected: HashMap<Uuid, Order> = HashMap::new();
        affected.insert(order.id, order.clone());

        if order.symbol != self.symbol {
            return (vec![], affected.into_values().collect());
        }

        let qty0 = parse_dec(&order.remaining).unwrap_or(Decimal::ZERO);
        if qty0 <= Decimal::ZERO {
            order.status = "rejected".to_string();
            order.updated_at = now;
            return (vec![], vec![order.clone()]);
        }

        let mut trades: Vec<Trade> = Vec::new();
        let mut executed_something = false;

        match order.side.as_str() {
            "buy" => {
                if order.type_ == "limit" && order.price.is_none() {
                    order.status = "rejected".to_string();
                    order.updated_at = now;
                    return (vec![], vec![order.clone()]);
                }
                while parse_dec(&order.remaining).unwrap_or(Decimal::ZERO) > Decimal::ZERO {
                    let best_ask = match self.best_ask_price() {
                        Some(p) => p,
                        None => break,
                    };
                    if order.type_ == "limit" {
                        let lim = parse_dec(order.price.as_ref().unwrap()).unwrap();
                        if best_ask > lim {
                            break;
                        }
                    }
                    let maker_id = {
                        let pl = self.ask_levels.get_mut(&best_ask).unwrap();
                        pl.top_valid(&self.orders)
                    };
                    let maker_id = match maker_id {
                        Some(id) => id,
                        None => break,
                    };
                    let maker_px = {
                        let m = self.orders.get(&maker_id).unwrap();
                        parse_dec(m.price.as_ref().unwrap()).unwrap()
                    };

                    let maker = self.orders.get_mut(&maker_id).unwrap();

                    let taker_rem = parse_dec(&order.remaining).unwrap();
                    let maker_rem = parse_dec(&maker.remaining).unwrap();
                    let exec_qty = min(taker_rem, maker_rem);
                    if exec_qty <= Decimal::ZERO {
                        break;
                    }
                    order.remaining = dec_str(taker_rem - exec_qty);
                    maker.remaining = dec_str(maker_rem - exec_qty);
                    order.updated_at = now;
                    maker.updated_at = now;
                    if parse_dec(&maker.remaining).unwrap_or(Decimal::ZERO) <= Decimal::ZERO {
                        maker.remaining = dec_str(Decimal::ZERO);
                        maker.status = "filled".to_string();
                        self.orders.remove(&maker_id);
                    } else {
                        maker.status = "partial_filled".to_string();
                    }
                    affected.insert(
                        maker_id,
                        self.orders
                            .get(&maker_id)
                            .cloned()
                            .unwrap_or_else(|| Order {
                                id: maker_id,
                                user_id: String::new(),
                                symbol: self.symbol.clone(),
                                side: "sell".to_string(),
                                type_: "limit".to_string(),
                                price: Some(dec_str(maker_px)),
                                quantity: String::new(),
                                remaining: dec_str(Decimal::ZERO),
                                status: "filled".to_string(),
                                created_at: now,
                                updated_at: now,
                                time_priority_seq: 0,
                            }),
                    );
                    executed_something = true;

                    trades.push(Trade {
                        id: Uuid::new_v4(),
                        symbol: self.symbol.clone(),
                        price: dec_str(maker_px),
                        quantity: dec_str(exec_qty),
                        buy_order_id: order.id,
                        sell_order_id: maker_id,
                        executed_at: now,
                    });
                }
                if parse_dec(&order.remaining).unwrap_or(Decimal::ZERO) <= Decimal::ZERO {
                    order.remaining = dec_str(Decimal::ZERO);
                    order.status = "filled".to_string();
                } else if order.type_ == "market" {
                    order.status = "partial_filled".to_string();
                } else {
                    if executed_something {
                        order.status = "partial_filled".to_string();
                    } else {
                        order.status = "open".to_string();
                    }
                    self.add_resting(order);
                    affected.insert(order.id, self.orders.get(&order.id).cloned().unwrap_or_else(|| order.clone()));
                }
            }
            "sell" => {
                if order.type_ == "limit" && order.price.is_none() {
                    order.status = "rejected".to_string();
                    order.updated_at = now;
                    return (vec![], vec![order.clone()]);
                }
                while parse_dec(&order.remaining).unwrap_or(Decimal::ZERO) > Decimal::ZERO {
                    let best_bid = match self.best_bid_price() {
                        Some(p) => p,
                        None => break,
                    };
                    if order.type_ == "limit" {
                        let lim = parse_dec(order.price.as_ref().unwrap()).unwrap();
                        if best_bid < lim {
                            break;
                        }
                    }
                    let maker_id = {
                        let pl = self.bid_levels.get_mut(&best_bid).unwrap();
                        pl.top_valid(&self.orders)
                    };
                    let maker_id = match maker_id {
                        Some(id) => id,
                        None => break,
                    };
                    let maker = self.orders.get_mut(&maker_id).unwrap();

                    let taker_rem = parse_dec(&order.remaining).unwrap();
                    let maker_rem = parse_dec(&maker.remaining).unwrap();
                    let exec_qty = min(taker_rem, maker_rem);
                    if exec_qty <= Decimal::ZERO {
                        break;
                    }
                    order.remaining = dec_str(taker_rem - exec_qty);
                    maker.remaining = dec_str(maker_rem - exec_qty);
                    order.updated_at = now;
                    maker.updated_at = now;

                    let maker_price_str = maker.price.clone();

                    if parse_dec(&maker.remaining).unwrap_or(Decimal::ZERO) <= Decimal::ZERO {
                        maker.remaining = dec_str(Decimal::ZERO);
                        maker.status = "filled".to_string();
                        self.orders.remove(&maker_id);
                    } else {
                        maker.status = "partial_filled".to_string();
                    }
                    affected.insert(
                        maker_id,
                        self.orders
                            .get(&maker_id)
                            .cloned()
                            .unwrap_or_else(|| Order {
                                id: maker_id,
                                user_id: String::new(),
                                symbol: self.symbol.clone(),
                                side: "buy".to_string(),
                                type_: "limit".to_string(),
                                price: maker_price_str.clone(),
                                quantity: String::new(),
                                remaining: dec_str(Decimal::ZERO),
                                status: "filled".to_string(),
                                created_at: now,
                                updated_at: now,
                                time_priority_seq: 0,
                            }),
                    );
                    executed_something = true;

                    let maker_px = parse_dec(maker_price_str.as_ref().unwrap()).unwrap();

                    trades.push(Trade {
                        id: Uuid::new_v4(),
                        symbol: self.symbol.clone(),
                        price: dec_str(maker_px),
                        quantity: dec_str(exec_qty),
                        buy_order_id: maker_id,
                        sell_order_id: order.id,
                        executed_at: now,
                    });
                }
                if parse_dec(&order.remaining).unwrap_or(Decimal::ZERO) <= Decimal::ZERO {
                    order.remaining = dec_str(Decimal::ZERO);
                    order.status = "filled".to_string();
                } else if order.type_ == "market" {
                    order.status = "partial_filled".to_string();
                } else {
                    if executed_something {
                        order.status = "partial_filled".to_string();
                    } else {
                        order.status = "open".to_string();
                    }
                    self.add_resting(order);
                    affected.insert(order.id, self.orders.get(&order.id).cloned().unwrap_or_else(|| order.clone()));
                }
            }
            _ => {
                order.status = "rejected".to_string();
                order.updated_at = now;
                return (vec![], vec![order.clone()]);
            }
        }

        self.bump_sequence();
        affected.insert(order.id, order.clone());
        let mut out: Vec<Order> = affected.into_values().collect();
        out.sort_by_key(|o| o.id);
        (trades, out)
    }

    pub fn cancel(&mut self, id: Uuid, now: DateTime<Utc>) -> Option<Order> {
        let o = self.orders.get(&id)?.clone();
        if o.status != "open" && o.status != "partial_filled" {
            return None;
        }
        let mut o = o;
        o.status = "cancelled".to_string();
        o.updated_at = now;
        self.orders.remove(&id);
        self.bump_sequence();
        Some(o)
    }

    pub fn snapshot(&mut self, depth: usize) -> BookSnapshot {
        let depth = if depth == 0 { 50 } else { depth };

        let mut bid_prices: Vec<Decimal> = self
            .bid_levels
            .iter()
            .filter(|(_, pl)| pl.total_remaining(&self.orders) > Decimal::ZERO)
            .map(|(p, _)| *p)
            .collect();
        bid_prices.sort_by(|a, b| b.cmp(a));

        let mut ask_prices: Vec<Decimal> = self
            .ask_levels
            .iter()
            .filter(|(_, pl)| pl.total_remaining(&self.orders) > Decimal::ZERO)
            .map(|(p, _)| *p)
            .collect();
        ask_prices.sort_by(|a, b| a.cmp(b));

        let best_bid = bid_prices.first().map(|p| dec_str(*p));
        let best_ask = ask_prices.first().map(|p| dec_str(*p));

        let mut bids = Vec::new();
        for p in bid_prices.into_iter().take(depth) {
            if let Some(pl) = self.bid_levels.get(&p) {
                let t = pl.total_remaining(&self.orders);
                if t > Decimal::ZERO {
                    bids.push(BookLevel {
                        price: dec_str(p),
                        quantity: dec_str(t),
                    });
                }
            }
        }
        let mut asks = Vec::new();
        for p in ask_prices.into_iter().take(depth) {
            if let Some(pl) = self.ask_levels.get(&p) {
                let t = pl.total_remaining(&self.orders);
                if t > Decimal::ZERO {
                    asks.push(BookLevel {
                        price: dec_str(p),
                        quantity: dec_str(t),
                    });
                }
            }
        }

        BookSnapshot {
            symbol: self.symbol.clone(),
            bids,
            asks,
            best_bid,
            best_ask,
            sequence: self.sequence,
        }
    }
}

#[derive(serde::Serialize)]
pub struct BookLevel {
    pub price: String,
    pub quantity: String,
}

#[derive(serde::Serialize)]
pub struct BookSnapshot {
    pub symbol: String,
    pub bids: Vec<BookLevel>,
    pub asks: Vec<BookLevel>,
    pub best_bid: Option<String>,
    pub best_ask: Option<String>,
    pub sequence: u64,
}
